library/file.lisp

```diff
@@ -0,0 +1,711 @@
+                                                   (IfExists/Supersede ':supersede))
+                                      :if-does-not-exist ':create))))
+
+  (declare %open-bidirectional (Pathname -> IFExists -> types:lisptype -> (Result FileError (FileStream :a))))
+  (define (%open-bidirectional path if-exists etype)
+    "Opens a two way stream for the given filepath and for a given type."
+    (lisp (Result FileError (FileStream :a)) (path if-exists etype)
+      (%handle-file-function (cl:open path
+                                      :direction :io
+                                      :element-type etype
+                                      :if-exists (cl:case if-exists
+                                                   (IfExists/ExistsError ':error)
+                                                   (IfExists/NewVersion ':new-version)
+                                                   (IfExists/Rename ':rename)
+                                                   (IfExists/RenameAndDelete ':rename-and-delete)
+                                                   (IfExists/Overwrite ':overwrite)
+                                                   (IfExists/Append ':append)
+                                                   (IfExists/Supersede ':supersede))
+                                      :if-does-not-exist ':create))))
+
+  (declare %open (StreamOptions -> types:lisptype -> (Result FileError (FileStream :a))))
+  (define (%open stream-options etype)
+    "Opens a FileStream for a given type and StreamOptions."
+    (match stream-options
+      ((Input path)
+       (%open-input path etype))
+      ((Output path exists)
+       (%open-output path exists etype))
+      ((Bidirectional path exists)
+       (%open-bidirectional path exists etype))))
+
+  ;;
+  ;; Other basic stream operations
+  ;;
+  (declare close ((FileStream :a) -> (Result FileError :b)))
+  (define (close stream)
+    "Closes a FileStream."
+    (lisp (Result FileError :a) (stream)
+      (%handle-file-function (cl:close stream))))
+
+  (declare abort ((FileStream :a) -> (Result FileError :b)))
+  (define (abort stream)
+    "Closes a FileStream and aborts all operations.."
+    (lisp (Result FileError :a) (stream)
+      (%handle-file-function (cl:close stream :abort cl:t))))
+
+  ;;
+  ;;
+  ;;
+
+  (declare read-char ((FileStream Char) -> (Result FileError Char)))
+  (define (read-char stream)
+    "Reads a character from an FileStream."
+    (lisp (Result FileError Char) (stream)
+      (cl:handler-case (Ok (cl:read-char stream))
+        (cl:end-of-file () (Err (EOF)))
+        (cl:error (c) (Err (LispError c))))))
+
+  (declare write-char ((FileStream Char) -> Char -> (Result FileError Unit)))
+  (define (write-char stream data)
+    "Writes a `Char` to the stream."
+    (lisp (Result FileError Unit) (stream data)
+      (%handle-file-function (cl:write-char data stream))))
+
+  (define-class (FileByte :a)
```

Comment:
Don't export this or `read-byte`/`write-byte`. I consider it an implementation detail.